snuba/query/processors/logical/granularity_processor.py

```diff
@@ -12,36 +31,155 @@
 DEFAULT_GRANULARITY_RAW = 60
 
 
-class GranularityProcessor(LogicalQueryProcessor):
-    """Use the granularity set on the query to filter on the granularity column"""
+class BaseGranularityProcessor(LogicalQueryProcessor):
+    @abstractmethod
+    def get_granularity(self, query: Query) -> Tuple[int, Callable[[Query, int], None]]:
+        raise NotImplementedError
 
-    @staticmethod
-    def __get_granularity(query: Query) -> int:
-        """Find the best fitting granularity for this query"""
-        requested_granularity = query.get_granularity()
+    @abstractmethod
+    def process_query(self, query: Query, query_settings: QuerySettings) -> None:
+        raise NotImplementedError
 
-        if requested_granularity is None:
-            return DEFAULT_GRANULARITY_RAW
-        elif requested_granularity > 0:
-            for granularity in reversed(GRANULARITIES_AVAILABLE):
-                if (requested_granularity % granularity) == 0:
+    @abstractmethod
+    def get_highest_common_available_granularity_multiple(
+        self,
+        selected_granularity: int,
+    ) -> int:
+        raise NotImplementedError
 
-                    return granularity
+    def find_granularity_in_expression(
+        self, expression: Optional[Expression]
+    ) -> Optional[int]:
+        match = FunctionCall(
+            String(ConditionFunctions.EQ),
+            (
+                Column(column_name=String("granularity")),
+                Literal(value=Param("granularity", Any(int))),
+            ),
+        ).match(expression)
+        if match is not None:
+            return match.integer("granularity")
 
-        raise InvalidGranularityException(
-            f"Granularity must be multiple of one of {GRANULARITIES_AVAILABLE}"
-        )
+        match = FunctionCall(
+            Param(
+                "operator",
+                Or([String(BooleanFunctions.AND), String(BooleanFunctions.OR)]),
+            ),
+            (Param("lhs", AnyExpression()), Param("rhs", AnyExpression())),
+        ).match(expression)
 
-    def process_query(self, query: Query, query_settings: QuerySettings) -> None:
-        granularity = self.__get_granularity(query)
+        if match is not None:
+            lhs_granularity = self.find_granularity_in_expression(
+                match.expression("lhs")
+            )
+            rhs_granularity = self.find_granularity_in_expression(
+                match.expression("rhs")
+            )
+            if lhs_granularity is None:
+                return rhs_granularity
+            elif rhs_granularity is None:
+                return lhs_granularity
+            else:
+                raise InvalidGranularityException(
```

Comment:
I don't think this is quite what the product team is looking for. Part of what they want to do is specify multiple granularities in a query, so they can do a single query that fetches a different granularity for different time periods. So something like

```
MATCH (...)
SELECT ...
WHERE (
    (granularity = 1 AND time > now() - INTERVAL 1 DAY)*
    OR (granularity = 2 AND time < now() - INTERVAL 1 DAY)
)
AND ...
```

In this example they can have the hourly granularity for the most recent 24 hours and the daily granularity for everything else. Note that they would still need top level time conditions that cover the entire range (`time >= ... AND time < ...`).

*Note that we don't actually support the INTERVAL syntax in SnQL, I was just using it to make the example easier to read.

Reply:
This PR isn't meant to support multiple granularities in a query. The objective is to first allow granularities to work inside the WHERE clause since this is currently broken. For the use-case you mentioned (supporting multiple granularities in a single query for different time periods), I think would build off of this fix and maybe should be described in a new ticket.

Reply:
Update the PR and its description to include this change. The use-case you describe is captured in this test https://github.com/getsentry/snuba/pull/4466/files#diff-0111cefeaef97e9d9b44d01013f69d4b529e1b1cbe1d9136e32268683f7e6311R449